src/coreclr/jit/optimizer.cpp

```diff
@@ -7067,53 +7153,113 @@ bool Compiler::optVNIsLoopInvariant(ValueNum vn, unsigned lnum, VNToBoolMap* loo
     return res;
 }
 
-/*****************************************************************************
- *
- *  Creates a pre-header block for the given loop - a preheader is a BBJ_NONE
- *  header. The pre-header will replace the current lpHead in the loop table.
- *  The loop has to be a do-while loop. Thus, all blocks dominated by lpHead
- *  will also be dominated by the loop-top, lpHead->bbNext.
- *
- */
-
+//------------------------------------------------------------------------------
+// fgCreateLoopPreHeader: Creates a pre-header block for the given loop.
+// A pre-header is a block outside the loop that falls through or branches to the loop
+// entry block. It is the only non-loop predecessor block to the entry block (thus, it
+// dominates the entry block). The pre-header replaces the current lpHead in the loop table.
+// The pre-header will be placed immediately before the loop top block, which is the first
+// block of the loop in program order.
+//
+// Once a loop has a pre-header, calling this function will immediately return without
+// creating another.
+//
+// If there already exists a block that meets the pre-header requirements, that block is marked
+// as a pre-header, and no flow graph modification is made.
+//
+// Note that the pre-header block can be in a different EH region from blocks in the loop, including the
+// entry block. Code doing hoisting is required to check the EH legality of hoisting to the pre-header
+// before doing so.
```

Comment:
Ah, I see you're allowing a wrong-EH-region existing block to serve as preheader. Seems like we shoud just create a new one in this case so we can perhaps hoist something.

Reply:
Reading below seems like you have thought of this ... wonder if we should check instead for (likely rare) case where the loop entry is the try entry and has backedges to it from outside the try. 

Or maybe it's not that rare given our limits on loop recognition?

Reply:
> wonder if we should check instead for (likely rare) case where the loop entry is the try entry and has backedges to it from outside the try.

I think I'm confused about where you propose this checking be done (or what it would restrict).

In general, the pre-header is put in the "correct" EH region for placing it as the block preceding the "top" block of the loop range. This might not be the same as the "top" block, nor the same as the "entry" block (for non-top-entry loops).

e.g., for loop:
```
    BB02 // pre-header, branches to BB04
    BB03 // loop "top", single-block "try" region 'A'
    BB04 <- entry block, and also the only block in a 'try' region 'B'
    BB05 // loop "bottom", branches to "top"
```
So, BB02 and BB03 and BB04 all have different try regions. If BB02/BB05 are in the same region (say, no try region), you could imagine hoisting something loop invariant from BB05 to BB02. You could also hoist from BB03 or BB04 to BB02 *if* it couldn't throw an exception.

Reply:
I guess all I'm really suggesting is that if there is an existing block that could serve as the preheader, and it's not in the same EH region as the loop, and the rest of the loop is all in one region, we might create a new block to be the preheader instead of adopting the one that's there.

Reply:
I added some code to check that the existing `head` has the same EH region of the pre-header we will create. If they are different, we continue and create a new pre-header; otherwise, we go ahead and use the current `head`. There were no SPMI diffs from this change.